torch/csrc/jit/runtime/profiling_graph_executor_impl.cpp

```diff
@@ -421,43 +421,37 @@ void ProfilingGraphExecutorImpl::runNoGradOptimizations(
     LowerSimpleTuples(graph);
     GRAPH_DEBUG("After LowerSimpleTuples\n", *graph);
 
+    // Remove prim::profile nodes and embed the profile info directly in the
+    // IR in value types. We're doing such transformation as optimizations
+    // that try to merge/fuse nodes in the graph (e.g. BatchMM and GraphFuser)
+    // work worse in the presence of intermittent prim::profile nodes.
+    // Optimizations relying on the type info are also responsible for
+    // inserting proper type checks. Once we're done with these optimizations
+    // we will wipe the tensor type information from the IR, so that it's not
+    // accidentally used by any other pass.
+    RemoveProfileNodesAndSpecializeTypes(graph);
+    GRAPH_DEBUG(
+        "After RemoveProfileNodesAndSpecializeTypes, before BatchMM\n", *graph);
+    // Rewrite subgraphs with many MMs into expressions that batch them.
+    BatchMM(graph);
```

Comment:
Thinking if this code was valid in the first place. `BatchMM` is a non-differentiable optimization but it should be okay I think since we should be guarding all nograd graphs w/ autodiff guards.

Reply:
BatchMM works fine with differentiability actually